Mathlib/Tactic/StacksAttribute.lean

```diff
@@ -43,48 +43,83 @@ It extends the `Tag` environment extension with the data `declName, tag, comment
 def addTagEntry {m : Type → Type} [MonadEnv m] (declName : Name) (tag comment : String) : m Unit :=
   modifyEnv (tagExt.addEntry · { declName := declName, tag := tag, comment := comment })
 
-/--
-The syntax for a Stacks tag: it is an optional number followed by an optional identifier.
-This allows `044Q3` and `GH3F6` as possibilities.
--/
-declare_syntax_cat stackTag
+open Parser
+
+/-- `stacksTag` is the node kind of Stacks Project Tags: a sequence of digits and
+uppercase letters. -/
+abbrev stacksTagKind : SyntaxNodeKind := `stacksTag
+
+/-- The main parser for Stacks Project Tags: it accepts any sequence of 4 digits or
+uppercase letters. -/
+def stacksTagFn : ParserFn := fun c s =>
+  let i := s.pos
+  let tag :=
+    Substring.mk c.input i c.input.endPos |>.takeWhile (fun c => c.isDigit || c.isUpper) |>.toString
```

Comment:
I think you should take here while the character is alphanumeric, then do a second check for this condition, so that you can distinguish 4 invalid characters from 3 valid ones.

Reply:
@eric-wieser can you spell out the branching logic you want here? Thanks.

Reply:
The `alphanum` was how it was in a previous iteration and I'm happy to restore it!

Reply:
I pushed my suggestion

Reply:
Yeah, I think that's an improvement.

Reply:
I like this and I am happy to see that the parser consumes any `alphanum` and then continues with further checks: I think that it really makes for a better user experience!